exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordIterator.java

```diff
@@ -0,0 +1,252 @@
+    if (lastBatchRead) {
+      return;
+    }
+    lastOutcome = outgoing.next(inputIndex, incoming);
+  }
+
+  public void mark() {
+    // Release all batches before current batch. [0 to startBatchPosition).
+    final Map<Range<Integer>,RecordBatchData> oldBatches = batches.subRangeMap(
+      Range.closedOpen(0, startBatchPosition)).asMapOfRanges();
+    for (Range<Integer> range : oldBatches.keySet()) {
+      oldBatches.get(range.lowerEndpoint()).clear();
+    }
+    batches.remove(Range.closedOpen(0, startBatchPosition));
+    markedInnerPosition = innerPosition;
+    markedOuterPosition = outerPosition;
+  }
+
+  public void reset() {
+    if (markedOuterPosition >= 0) {
+      // Move to rbd for markedOuterPosition.
+      final RecordBatchData rbdNew = batches.get(markedOuterPosition);
+      final RecordBatchData rbdOld = batches.get(startBatchPosition);
+      Preconditions.checkArgument(rbdOld != null);
+      Preconditions.checkArgument(rbdNew != null);
+      if (rbdNew != rbdOld) {
+        container.transferOut(rbdOld.getContainer());
+        container.transferIn(rbdNew.getContainer());
+      }
+      innerPosition = markedInnerPosition;
+      outerPosition = markedOuterPosition;
+      startBatchPosition = batches.getEntry(outerPosition).getKey().lowerEndpoint();
+      innerRecordCount = batches.getEntry(outerPosition).getKey().upperEndpoint() - startBatchPosition;
+      markedInnerPosition = markedOuterPosition = -1;
+    }
+  }
+
+  // TODO: Experimental: Move forward by delta (may cross one or more record batches)
+  public void forward(int delta) {
+    Preconditions.checkArgument(delta >= 0);
+    Preconditions.checkArgument(delta + outerPosition < totalRecordCount);
+    final int nextOuterPosition = delta + outerPosition;
+    final RecordBatchData rbdNew = batches.get(nextOuterPosition);
+    final RecordBatchData rbdOld = batches.get(outerPosition);
+    Preconditions.checkArgument(rbdNew != null);
+    Preconditions.checkArgument(rbdOld != null);
+    container.transferOut(rbdOld.getContainer());
+    // Get vectors from new position.
+    container.transferIn(rbdNew.getContainer());
+    outerPosition = nextOuterPosition;
+    startBatchPosition = batches.getEntry(outerPosition).getKey().lowerEndpoint();
+    innerPosition = outerPosition - startBatchPosition;
+    innerRecordCount = batches.getEntry(outerPosition).getKey().upperEndpoint() - startBatchPosition;
+  }
+
+  public IterOutcome next() {
+    if (finished()) {
+      return lastOutcome;
+    }
+    final int nextOuterPosition = outerPosition + 1;
+    final int nextInnerPosition = innerPosition + 1;
+    if (!initialized || nextOuterPosition >= totalRecordCount) {
+      nextBatch();
+      switch (lastOutcome) {
+        case NONE:
+        case STOP:
+          // No more data, disallow reads unless reset is called.
+          outerPosition = nextOuterPosition;
+          lastBatchRead = true;
+          break;
+        case OK_NEW_SCHEMA:
```

Comment:
In the case of ok new schema, you should remove vectors that have been removed from upstream.